### PR TITLE
Fix auto fermion system for non-QN tensors

### LINKED ISSUE #1002: [ITensors] [BUG] StackOverflowError when using auto_fermion
The REPL raises a StackOverflowError when initializing an MPS over fermionic sites, while having auto_fermion active. For example the code 

```julia
using ITensors

ITensors.enable_auto_fermion()

n = 10
s = siteinds("Electron", n)
psi = randomMPS(s, j -> isodd(j) ? "Up" : "Dn", linkdims=20)
```

returns the following error 

```julia
ERROR: StackOverflowError:
Stacktrace:
 [1] has_fermionic_subspaces(T::Tuple{Index{Int64}, Index{Int64}}) (repeats 2 times)
   @ ITensors ~/.julia/packages/ITensors/PisGD/src/physics/fermions.jl:46
```

This happen also when using the ```"Fermion"``` SiteType. The error is raised every time an MPS is manipulated after enabling auto_fermion, so for example 

```julia 

using ITensors

n = 10
s = siteinds("Fermion", n)
psi = randomMPS(s, j -> isodd(j) ? "1" : "0", linkdims=20)

ITensors.enable_auto_fermion()

inner(psi',psi)
```

returns the same StackOverflow error.

 - Output from `versioninfo()`:
```julia
julia> versioninfo()
Julia Version 1.8.0
Commit 5544a0fab76 (2022-08-17 13:38 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.3.0)
  CPU: 10 × Apple M1 Pro
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
  Threads: 8 on 8 virtual cores
Environment:
  JULIA_EDITOR = code
  JULIA_NUM_THREADS = 8
```
 - Output from `using Pkg; Pkg.status("ITensors")`:
```julia
julia> using Pkg; Pkg.status("ITensors")
Status `~/.julia/environments/v1.8/Project.toml`
  [9136182c] ITensors v0.3.20
```

### PR DESCRIPTION
**Pull request by emstoudenmire**
# Description

Fixes #1002, by removing an ambiguous (or self-calling) function overload and replacing it with some more constrained versions.

# How Has This Been Tested?

Added a small new test that reproduces the issue in #1002.

# Checklist:

- [X] My code follows the style guidelines of this project. Please run `using JuliaFormatter; format(".")` in the base directory of the repository (`~/.julia/dev/ITensors`) to format your code according to our style guidelines.
- [X] I have performed a self-review of my own code.
- [X] I have commented my code, particularly in hard-to-understand areas.
- [X] I have added tests that verify the behavior of the changes I made.
- [X] I have made corresponding changes to the documentation.
- [X] My changes generate no new warnings.
- [X] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: src/physics/fermions.jl [hunk 1]
```diff
@@ -34,7 +34,7 @@ end
 
 isfermionic(i::Index) = has_fermionic_subspaces(i)
 
-has_fermionic_subspaces(is::IndexSet) = false
+has_fermionic_subspaces(is::Union{IndexSet,NTuple{N,Index}}) where {N} = false
```
Comment (mtfishman) [inline src/physics/fermions.jl:37]: `IndexSet` was deprecated a while ago, you could use `Indices` here:
https://github.com/ITensor/ITensors.jl/blob/3ac68593e13d383d9a50ccf386aee15a73dcbaea/src/indexset.jl#L24

File: src/physics/fermions.jl [hunk 2]
```diff
@@ -34,7 +34,7 @@ end
...
 
 function has_fermionic_subspaces(is::Union{QNIndexSet,NTuple{N,QNIndex}}) where {N}
```
Comment (mtfishman) [inline src/physics/fermions.jl:39]: You could use `QNIndices` here:
https://github.com/ITensor/ITensors.jl/blob/23bc031e45d1fac0bbc839e524f85c740ac48b66/src/qn/qnindexset.jl#L4

File: src/physics/fermions.jl (hunk 1 above)
Comment (emstoudenmire) [reply src/physics/fermions.jl:37]: Good to know – changing it now thanks

File: src/physics/fermions.jl (hunk 2 above)
Comment (emstoudenmire) [reply src/physics/fermions.jl:39]: Great

**Pull request merged**